### ISSUE TITLE
Restore the ability to draw multiple samples with Open Source models

### ISSUE THREAD

**Issue by rlouf** [outlines-dev/outlines#416]
This was removed in #366 to simplify the PR, and should be added again. This will require to be careful with the shape and an added dimension will need to be added to take sample shape into account. The mechanism implemented there can be re-used when implementing Beam Search #258

**Comment by lapp0** [comment]
I'm looking to implement beam search.

I'm wondering whether we could simply use the sequence tokens as the ID of the sequence?

https://github.com/outlines-dev/outlines/commit/8b1ff9a7863915f5fdee421d4cae8f0840d58b33#diff-f65ffb5f52b2e358c713ccb8f32a700769426c6c8b655f689e3cdccae07d22ac

For 1000 token sequences, I can generate 25,000 keys on my machine per second, so it shouldn't be a substantial bottleneck.

**Comment by rlouf** [comment]
My understanding is that vLLM needs sequence ids because they're doing continuous batching, and we wouldn't need to assign an id to sequences here.

I'm still hesitating between using one big tensor of shape `n_samples x n_batch x n_token_ids` like we did before the refactor, or breaking the sequences down like this:

```python
class Generation:
	sequences: List[Sequence]
	

class Sequence:
	prompt_token_ids: List[int]
	generated_token_ids: List[int]
	logprob: float
```

For beam search, a `Sequence` would be a beam and `Generation` correspond to an input prompt. This is the way vLLM does it, and they create new tensors at each step. What do you think?

**Comment by lapp0** [comment]
I don't think I highlighted the component I want to address properly.

# Efficient Caching

In the existing outlines codebase, the `RegexLogitsProcessor` is called with seq_id which corresponds to a distinct FSM state https://github.com/outlines-dev/outlines/blob/0355ab4272a5d7e4d94c4a53a52593f885b81a61/outlines/serve/vllm.py#L68-L70

`_apply_logits_processors` is patched in outlines to pass seq_id.

This works, but if we used `tuple(input_ids)` as a distinct key, we would be able to prevent repeated work for a given `(fsm, input_ids)`

Additionally, we would no longer need to patch `_patched_apply_logits_processors` which would break a dependency on specifics of vLLMs codebase, and allow users to use other logits processors as well.

# Multiple Sampling

I'm not fluent in outlines codebase, so if you could point out what I'm missing I'd appreciate it. 

My understanding is that outlines allows vLLMs AsyncEngine to handle all KV management and simply provides a logits processor https://github.com/outlines-dev/outlines/blob/0355ab4272a5d7e4d94c4a53a52593f885b81a61/outlines/serve/serve.py#L80

This is a clean paradigm and it makes sense to me that outlines would leave KV cache management to `transformers` as well, and simply provide an efficient logits processor.

Regarding `outlines/outlines/generate/api.py`, why wouldn't we be able to simplify `SequenceGenerator` so it allows `transformers` to do the bulk of the work in managing generation? It appears that we would be able to provide `transformers` with a logits processor which retrieves the cached FSM state as `fn(generation_token_ids)` rather than tracking and manipulating `FSMState` within `SequenceGenerator`.

**Comment by rlouf** [comment]
Would you mind copy/pasting your comment into a discussion and I'll answer there? Just so we stay on topic here and your comment is easier to find for future readers.

**Comment by lapp0** [comment]
https://github.com/outlines-dev/outlines/discussions/501

**Comment by rlouf** [comment]
Here are the changes that need to be implemented in order to restore the ability to generate several samples for each sequence:

1. Update `init_generator_state` to it expands the prompt ids from `(n_batches, n_tokens)` to `(n_samples, n_batches, n_tokens)` by duplicating the prompts. A single prompt with 10 tokens would lead to `token_ids` with shape `(1, 1, 10)`, a batch of 3 prompts to a shape `(1, 3, 10)` (with padding) and 7 samples for a batch of 3 prompts with shape `(7, 3, 10)`. We keep singleton dimensions to simplify the code.
2. Reshape the `token_ids` array to `(n_batches * n_samples, n_tokens)` before calling the token generator
3. Decode the result.
4. Reshape the decoded sequences to `(n_samples, n_batches, n_tokens)`

We will need to add tests for `init_generator_state`, `sequence_generator` and the sampling algorithms.
